planner/core/rule_max_min_eliminate.go

```diff
@@ -18,71 +18,188 @@ import (
 	"github.com/pingcap/parser/ast"
 	"github.com/pingcap/parser/mysql"
 	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
 	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/ranger"
 )
 
 // maxMinEliminator tries to eliminate max/min aggregate function.
 // For SQL like `select max(id) from t;`, we could optimize it to `select max(id) from (select id from t order by id desc limit 1 where id is not null) t;`.
 // For SQL like `select min(id) from t;`, we could optimize it to `select max(id) from (select id from t order by id limit 1 where id is not null) t;`.
+// For SQL like `select max(id), min(id) from t;`, we could optimize it to the cartesianJoin result of the two queries above if `id` has an index.
 type maxMinEliminator struct {
 }
 
 func (a *maxMinEliminator) optimize(ctx context.Context, p LogicalPlan) (LogicalPlan, error) {
-	a.eliminateMaxMin(p)
-	return p, nil
+	return a.eliminateMaxMin(p), nil
 }
 
-// Try to convert max/min to Limit+Sort operators.
-func (a *maxMinEliminator) eliminateMaxMin(p LogicalPlan) {
-	// We don't need to guarantee that the child of it is a data source. This transformation won't be worse than previous.
-	if agg, ok := p.(*LogicalAggregation); ok {
-		// We only consider case with single max/min function.
-		if len(agg.AggFuncs) != 1 || len(agg.GroupByItems) != 0 {
-			return
+// Compose the scalar aggregations by cartesianJoin.
+func (a *maxMinEliminator) composeAggsByInnerJoin(aggs []*LogicalAggregation) (plan LogicalPlan) {
+	plan = aggs[0]
+	sctx := plan.SCtx()
+	for i := 1; i < len(aggs); i++ {
+		join := LogicalJoin{JoinType: InnerJoin}.Init(sctx)
+		join.SetChildren(plan, aggs[i])
+		join.schema = buildLogicalJoinSchema(InnerJoin, join)
+		join.cartesianJoin = true
+		plan = join
+	}
+	return
+}
+
+// checkColCanUseIndex checks the following conditions:
+// 1. whether the col is the prefix of an index.
+// 2. whether all of the selection conditions can be pushed down to the index range.
+func (a *maxMinEliminator) checkColCanUseIndex(plan LogicalPlan, col *expression.Column) bool {
+	switch p := plan.(type) {
+	case *LogicalSelection:
+		// Check whether all of the conditions can be pushed down as accessConds.
+		if _, filterConds := ranger.DetachCondsForColumn(p.ctx, p.Conditions, col); len(filterConds) != 0 {
+			return false
 		}
-		f := agg.AggFuncs[0]
-		if f.Name != ast.AggFuncMax && f.Name != ast.AggFuncMin {
-			return
+		return a.checkColCanUseIndex(p.children[0], col)
+	case *DataSource:
+		// Check whether there is an accessPath can use index for col.
+		for _, path := range p.possibleAccessPaths {
+			if path.isTablePath {
+				if col == p.handleCol {
+					return true
+				}
+			} else {
+				if col.ColName.L == path.index.Columns[0].Name.L {
```

Comment:
You may need to add a case for prefix index.
```sql
mysql> create table t(a char(50), b int, index idx(a(10)));
Query OK, 0 rows affected (0.01 sec)

mysql> explain select max(a), min(a) from t;
+--------------------------------+----------+------+------------------------------------------------------------+
| id                             | count    | task | operator info                                              |
+--------------------------------+----------+------+------------------------------------------------------------+
| HashLeftJoin_16                | 1.00     | root | CARTESIAN inner join, inner:StreamAgg_49                   |
| ├─StreamAgg_22                 | 1.00     | root | funcs:max(test.t.a)                                        |
| │ └─TopN_23                    | 1.00     | root | test.t.a:desc, offset:0, count:1                           |
| │   └─TableReader_32           | 1.00     | root | data:TopN_31                                               |
| │     └─TopN_31                | 1.00     | cop  | test.t.a:desc, offset:0, count:1                           |
| │       └─Selection_30         | 9990.00  | cop  | not(isnull(test.t.a))                                      |
| │         └─TableScan_29       | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
| └─StreamAgg_49                 | 1.00     | root | funcs:min(test.t.a)                                        |
|   └─TopN_50                    | 1.00     | root | test.t.a:asc, offset:0, count:1                            |
|     └─TableReader_57           | 1.00     | root | data:TopN_56                                               |
|       └─TopN_56                | 1.00     | cop  | test.t.a:asc, offset:0, count:1                            |
+--------------------------------+----------+------+------------------------------------------------------------+
11 rows in set (0.00 sec)
```

Reply:
addressed